Feat/course registrations

frooooooo7 opened the pull request.

frooooooo7 committed:
feat: create CourseRegistrationTable, model and controller

frooooooo7 committed:
PHP Linting (Pint)

frooooooo7 committed:
feat: implement index and update methods for course registrations

frooooooo7 committed:
PHP Linting (Pint)

frooooooo7 committed:
refactor: moved data handling logic from service to value objects

frooooooo7 committed:
PHP Linting (Pint)

marioooo0o commented on database/migrations/2025_01_08_183400_create_course_registrations_table.php:
```diff
@@ -0,0 +1,30 @@
+            $table->id();
+            $table->foreignId('course_id')->constrained();
+            $table->foreignId('user_id')->constrained();
+            $table->string('status', 10);
```
https://laravel.com/docs/11.x/migrations#column-method-enum

marioooo0o commented on routes/api.php:
```diff
@@ -33,3 +34,12 @@
+Route::middleware('auth:sanctum')->group(function () {
+    Route::controller(CourseRegistrationController::class)->group(function () {
+        Route::get('course-registrations', 'index');
+        Route::post('course-registrations/{courseId}/create', 'store');
+        Route::put('course-registrations/{registrationId}', 'update');
```
stosuj model binding, dla post podaj course_id w body, a dla put model binding unikniesz wtedy potrzeby pamiętania o sprawdzaniu findOrFail i dane będą bardziej hermetyczne, będziesz przekazywał do serwisu obiekt klasy modelu Course/CourseRegistration, zamiast id i nie będzie potrzeby sprawdzać za każdym razem czy jest taki obiekt w bazie danych

marioooo0o commented on app/Http/Requests/UpdateCourseRegistrationRequest.php:
```diff
@@ -0,0 +1,41 @@
+            'status' => [
+                'required',
+                'string',
+                'in:'.implode(',', [RegistrationStatus::ACCEPTED->value, RegistrationStatus::REJECTED->value]),
```
Tutaj można poza hard codowaniem wartości, stworzyć metodę w Enumie zwracająca te opcje które są dostępne a następnie coś w stylu:
Rule::in(array_column(RegistrationStatus::nazwa_metody(), 'value'))],

frooooooo7 committed:
refactor: use enum values in migration

frooooooo7 committed:
refactor:
- model binding for CourseRegistration in update method
- add values() method to RegistrationStatus enum and add Rule:: in validation for status

frooooooo7 committed:
feat: add seeder for CourseRegistrations

frooooooo7 committed:
PHP Linting (Pint)

marioooo0o commented on routes/api.php:
```diff
@@ -33,3 +34,12 @@
 Route::get('/google/callback', [GoogleAuthController::class, 'handleGoogleCallback']);
+
+// CourseRegistrations
+Route::middleware('auth:sanctum')->group(function () {
+    Route::controller(CourseRegistrationController::class)->group(function () {
+        Route::get('course-registrations', 'index');
+        Route::post('course-registrations', 'store');
+        Route::put('course-registrations/{courseRegistration}', 'update');
+    });
+});
```
zastosuj model binding czyli zrób trasę /courses/{course}/registration

i wtedy we wszyskich trasach będzie widomo do którego kursu się to tyczy, unikniemy podawania go w body
https://laravel.com/docs/11.x/controllers#restful-nested-resources

marioooo0o reviewed (changes requested):
plus te poprawki z enum

frooooooo7 committed:
refactor: CourseRegistration - enum, migration, requests and routes

frooooooo7 committed:
PHP Linting (Pint)

frooooooo7 committed:
refactor: split registration functionality between StudentCourseRegistration and AdminCourseRegistration
feat: add StudentCourseRegistrationController and its functionalities

frooooooo7 committed:
PHP Linting (Pint)

frooooooo7 committed:
refactor:
- rename Student to Guest
- add store function to AdminCourseRegistrationController

frooooooo7 committed:
PHP Linting (Pint)

frooooooo7 committed:
fix: store properly registartions to course_user

frooooooo7 committed:
PHP Linting (Pint)

frooooooo7 committed:
Merge branch 'dev' into feat/course-registrations

frooooooo7 committed:
feat: add endpoint to retrieve list of students

frooooooo7 committed:
PHP Linting (Pint)

frooooooo7 committed:
feat: add functionality to accept and decline course registrations

frooooooo7 committed:
PHP Linting (Pint)

marioooo0o commented on app/Http/Controllers/CourseRegistration/GuestCourseRegistrationController.php:
```diff
@@ -0,0 +1,19 @@
+{
+    public function __construct(protected GuestCourseRegistrationService $guestCourseRegistrationService) {}
+
+    public function store(GuestCourseRegistrationRequest $request, Course $course): GuestCourseRegistrationResource
+    {
+        return new GuestCourseRegistrationResource($this->guestCourseRegistrationService->store($request->getGuestCourseRegistration($course)));
+    }
```
![image](https://github.com/user-attachments/assets/b3705e7c-a23d-401d-902a-4ad450f15248)
brak konsekwensji raz wracasz resource w store a raz tylko informację

marioooo0o commented on app/Http/Requests/CourseRegistration/GuestCourseRegistrationRequest.php:
```diff
@@ -0,0 +1,62 @@
+                'max:30',
+                'regex:/^[a-zA-ZÀ-ž\s\'-]+$/',
+            ],
+            'email' => [
```
mail w bazie jest unique

frooooooo7 committed:
feat: implement search by name and last name for drivers

frooooooo7 committed:
refactor: requested changes

frooooooo7 committed:
PHP Linting (Pint)

marioooo0o commented on app/Http/Requests/CourseRegistration/GuestCourseRegistrationRequest.php:
```diff
@@ -24,28 +24,10 @@ public function authorize(): bool
+            'name' => 'required|string|min:2|max:30|regex:/^[a-zA-ZÀ-ž\s\'-]+$/',
+            'last_name' => 'required|string|min:2|max:30|regex:/^[a-zA-ZÀ-ž\s\'-]+$/',
+            'email' => 'required|email|unique:users,email',
+            'phone' => 'required|regex:/^\+?\d{9,15}$/',
```
zapoznaj się z  tym:
https://laravel-news.com/laravel-phone

spróbujmy tego

marioooo0o commented on app/Services/Admin/AdminStudentService.php:
zrób to tak jak mówilismy podobnie do kursów wraz z paginacją tak jak tam

frooooooo7 committed:
refactor: use QueryBuilder instead of AdminStudentService in students index

frooooooo7 committed:
PHP Linting (Pint)

frooooooo7 committed:
refactor: improve phone number validation and database storage

frooooooo7 committed:
PHP Linting (Pint)

frooooooo7 committed:
refactor: change the returned value of phone number in UserResource

frooooooo7 committed:
PHP Linting (Pint)
